Add Fabric implementation for Android

zolbooo opened the pull request:
Hello, thank you for work on this project 👋

I have updated example app to use React Native `0.69.0-rc.6`, which is the first release to enable new architecture by default. Also I have added implementation of VectorDrawable component for Fabric. I have build application with and without new arch enabled and example app seems to work correctly.

Test plan:
- [x] Android - Fabric example app
- [x] Android - Paper example app (built with `newArchEnabled=false` in `gradle.properties`
- [x] iOS - Fabric app (`USE_FABRIC=1 RCT_NEW_ARCH_ENABLED=1 pod install`)
- [x] iOS - Paper app (Regular `pod install`)
- [ ] Some props could be implemented incorrectly because `RNVectorDrawableNativeComponent` and `index.d.ts` typings differ

Muugii-meta committed:
build: add react-native typings

Muugii-meta committed:
build(example): update dependencies

Muugii-meta committed:
build(example-android): update gradle

Muugii-meta committed:
build(example-android): update AGP

Muugii-meta committed:
build(example): enable react-native gradle plugin with codegen

Muugii-meta committed:
build(example): use gradle 7.3.3

Muugii-meta committed:
build(example): use implementation instead of configurations.compile syntax

Muugii-meta committed:
build(android): enable new architecture

Muugii-meta committed:
build: add options for new architecture

Muugii-meta committed:
build(example): bump build tools version

Muugii-meta committed:
build(example-android): add JNI configuration

Muugii-meta committed:
build(example-app): add packagingOptions

Muugii-meta committed:
feat(android-example): configure TurboModules+Fabric app

Muugii-meta committed:
build(example): update compile SDK version

Muugii-meta committed:
fix(example-app): missing mReactNativeHost declaration

Muugii-meta committed:
fix(example-app): package name and missing BuildConfig import

Muugii-meta committed:
fix(example-app): remove non-existing library

Muugii-meta committed:
fix(example-app): JNI build makefile

Muugii-meta committed:
build(example-app): update repositories

Muugii-meta committed:
build(example-app): update flipper

Muugii-meta committed:
feat: add VectorDrawable Fabric component definition

Muugii-meta committed:
feat(vector-drawable): add fabric and paper implementations

Muugii-meta committed:
fix(example-app): add missing JNI code

Muugii-meta committed:
fix(gradle-config): define source directories properly

Muugii-meta committed:
feat(fabric): initialize components registry automatically

Muugii-meta committed:
fix(paper): add missing import for ThemedReactContext

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/klarna-incubator/react-native-vector-drawable?pullRequest=5) <br/>All committers have signed the CLA.

Muugii-meta committed:
build(ios): update pods, increase target platform version

Muugii-meta committed:
feat(ios): add Fabric AppDelegate implementation

Muugii-meta committed:
chore(ios): add IDEWorkspaceChecks.plist

Muugii-meta committed:
build(ios): use c++17

zolbooo commented:
Could you review this, @oblador?

oblador commented on android/build.gradle:
```diff
@@ -23,6 +30,46 @@ android {
+                            "NDK_TOOLCHAIN_VERSION=clang",
+                            "GENERATED_SRC_DIR=${appProject.buildDir}/generated/source",
+                            "PROJECT_BUILD_DIR=${appProject.buildDir}",
+                            "REACT_ANDROID_DIR=${appProject.rootDir}/../node_modules/react-native/ReactAndroid",
```
This is an unsafe assumption about where react-native package is located that works with this project, but breaks mono-repos (that we use at Klarna). See this [gradle file](https://github.com/oblador/react-native-performance/blob/master/packages/react-native-performance/android/build.gradle#L15-L28) for an example on how that can be done in a more dynamic fashion. I've discussed this with Meta but no better solution in place than this from what I know.

oblador commented on android/build.gradle:
```diff
@@ -23,6 +30,46 @@ android {
+        }
+    }
+    packagingOptions {
+        // See: https://github.com/software-mansion/react-native-screens/blob/main/android/build.gradle#L90
```
Weird, but ok

oblador commented on android/build.gradle:
```diff
@@ -32,18 +79,24 @@ android {
 }
 
 repositories {
-    if (project == rootProject) {
```
Any particular reason why you changed this setup?

oblador commented on android/build.gradle:
```diff
@@ -32,18 +79,24 @@ android {
 }
+
+if (isNewArchitectureEnabled()) {
+    react {
```
For monorepo setups I believe you need to pass reactNativeDir/codegenDir too as in the linked example

oblador commented on android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableManager.java:
```diff
@@ -0,0 +1,61 @@
+
+@ReactModule(name = VectorDrawableManagerImpl.NAME)
+public class VectorDrawableManager extends SimpleViewManager<ImageView>
+		implements RNVectorDrawableManagerInterface<ImageView> {
```
Mixing of spaces and tab indentation in this file, please make sure all java files are using 4 spaces

oblador commented on android/src/main/jni/VectorDrawableComponentsRegistry.cpp:
```diff
@@ -0,0 +1,54 @@
+
+void VectorDrawableComponentsRegistry::registerNatives() {
+  registerHybrid({
+      makeNativeMethod("initHybrid", VectorDrawableComponentsRegistry::initHybrid),
```
```suggestion
    makeNativeMethod("initHybrid", VectorDrawableComponentsRegistry::initHybrid),
```

oblador commented on android/src/main/jni/VectorDrawableComponentsRegistry.h:
```diff
@@ -0,0 +1,34 @@
+};
+
+} // namespace react
+} // namespace facebook
```
```suggestion
} // namespace facebook

```

oblador commented on android/src/main/jni/VectorDrawableComponentsRegistry.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+} // namespace react
+} // namespace facebook
```
```suggestion
} // namespace facebook

```

oblador commented on android/src/main/jni/OnInit.cpp:
```diff
@@ -0,0 +1,9 @@
+  return facebook::jni::initialize(vm, [] {
+    facebook::react::VectorDrawableComponentsRegistry::registerNatives();
+  });
+}
```
```suggestion
}

```

oblador commented on android/src/main/jni/Android.mk:
```diff
@@ -0,0 +1,37 @@
+
+LOCAL_CFLAGS := -DLOG_TAG=\"ReactNative\" -fexceptions -frtti -std=c++17 -Wall
+
+include $(BUILD_SHARED_LIBRARY)
```
```suggestion
include $(BUILD_SHARED_LIBRARY)

```

oblador commented on android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableComponentsRegistry.java:
```diff
@@ -0,0 +1,28 @@
+  public static VectorDrawableComponentsRegistry register(ComponentFactory componentFactory) {
+    return new VectorDrawableComponentsRegistry(componentFactory);
+  }
+}
```
For the pedantic smurf inside me, please always have an empty new line at the end of the file 🙏 
```suggestion
}

```

oblador commented on example/package.json:
```diff
@@ -11,9 +11,11 @@
+    "@klarna/react-native-vector-drawable": "*",
+    "react": "18.0.0",
+    "react-native": "0.69.0-rc.6",
+    "react-native-codegen": "^0.69.1",
```
I didn't think it was necessary to explicitly include codegen/gradle-plugin dependencies

oblador commented on src/RNVectorDrawableNativeComponent.ts:
```diff
@@ -0,0 +1,31 @@
+} from 'react-native';
+
+// @ts-ignore
+import {WithDefault} from 'react-native/Libraries/Types/CodegenTypes';
```
```suggestion
import { WithDefault } from 'react-native/Libraries/Types/CodegenTypes';
```

oblador reviewed (changes requested):
Absolutely fantastic PR! Added a bunch of comments, mostly minor, but I'm missing updated README instructions and maybe a mention of Fabric support.

oblador commented:
@zolbooo I'll be going on vacation tomorrow just as a heads up. 👍

zolbooo commented:
Okay, I will push my commits in a few hours.

Muugii-meta committed:
fix(android-build): find node_modules location dynamically

Hardcoding REACT_ANDROID_DIR and REACT_ANDROID_BUILD_DIR could break monorepo builds,
use findNodeModules to locate react-native in dynamic fashion.

Muugii-meta committed:
docs(android-build): add note on packagingOptions config

Muugii-meta committed:
fix(android-build): use repositories only if it's root project

Muugii-meta committed:
fix(android-build): add missing reactNativeDir and codegenDir options

Muugii-meta committed:
style(java): use 4 spaces for intendation

Muugii-meta committed:
style: add missing newlines at the end of files

Muugii-meta committed:
build(example): remove unnecessary react-native-codegen dependency

Muugii-meta committed:
docs(readme): update copyright year

Muugii-meta committed:
build(example): update react-native to stable version

Muugii-meta committed:
docs(readme): add Fabric installation section

Muugii-meta committed:
style(ts): format named import properly

Muugii-meta committed:
fix(build-gradle): add missing import

zolbooo commented:
Everything seems good to me now, please check 🚀

oblador commented:
Back from vacation now :D I tried to run the example project with fabric turned off (hence using paper), but it failed to compile with this error:
```
node_modules/@klarna/react-native-vector-drawable/android/src/paper/java/com/klarna/vectordrawable/VectorDrawableManager.java:22: error: invalid method declaration; return type required
    public ColoredViewManager(ReactApplicationContext reactContext) {
           ^
1 error
```

I didn't look much further into this error, but I'm guessing it's because codegen not being run in old architecture.

Muugii-meta committed:
fix(paper): VectorDrawableManager constructor name

zolbooo commented:
> Back from vacation now :D I tried to run the example project with fabric turned off (hence using paper), but it failed to compile with this error:
> 
> ```
> 
> node_modules/@klarna/react-native-vector-drawable/android/src/paper/java/com/klarna/vectordrawable/VectorDrawableManager.java:22: error: invalid method declaration; return type required
> 
>     public ColoredViewManager(ReactApplicationContext reactContext) {
> 
>            ^
> 
> 1 error
> 
> ```
> 
> 
> 
> I didn't look much further into this error, but I'm guessing it's because codegen not being run in old architecture. 

Oops, I forgot to rename constructor in Paper implementation. Please check the latest commit.

oblador commented:
Still same error for me 🤔 Did you manage to run this locally?

Muugii-meta committed:
feat(gradle): use ndk path and version from root project

Muugii-meta committed:
fix(paper): add missing imports

Muugii-meta committed:
feat(paper-impl): add @NonNull annotations

Muugii-meta committed:
fix(fabric-impl): add missing @NonNull annotation

zolbooo commented:
My bad, I was editing `node_modules` in example instead of editing library files directly. Now it compiles on both archtitectures. Also added support for M1 devices in [4b47a91](https://github.com/klarna-incubator/react-native-vector-drawable/pull/5/commits/4b47a91d9df47057fb36c954323f60a7bb1c6e64).
